realm/realm-library/src/androidTest/java/io/realm/internal/RealmNotifierTests.java

```diff
@@ -0,0 +1,198 @@
+ * limitations under the License.
+ */
+package io.realm.internal;
+
+
+import android.support.test.runner.AndroidJUnit4;
+
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+
+import java.util.concurrent.atomic.AtomicInteger;
+
+import io.realm.RealmChangeListener;
+import io.realm.RealmConfiguration;
+import io.realm.internal.android.AndroidRealmNotifier;
+import io.realm.rule.RunInLooperThread;
+import io.realm.rule.RunTestInLooperThread;
+import io.realm.rule.TestRealmConfigurationFactory;
+
+import static junit.framework.Assert.assertEquals;
+import static junit.framework.Assert.fail;
+
+@RunWith(AndroidJUnit4.class)
+public class RealmNotifierTests {
+    @Rule
+    public final TestRealmConfigurationFactory configFactory = new TestRealmConfigurationFactory();
+    @Rule
+    public final RunInLooperThread looperThread = new RunInLooperThread();
+
+    private RealmConfiguration config;
+    Capabilities capabilitiesCanDeliver = new Capabilities() {
+        @Override
+        public boolean canDeliverNotification() {
+            return true;
+        }
+
+        @Override
+        public void checkCanDeliverNotification(String exceptionMessage) {
+        }
+    };
+
+    @Before
+    public void setUp() throws Exception {
+        config = configFactory.createConfiguration();
+    }
+
+    @After
+    public void tearDown() {
+    }
+
+    private SharedRealm getSharedRealm(RealmConfiguration config) {
+        return SharedRealm.getInstance(config, null, true);
+    }
+
+    @Test
+    @RunTestInLooperThread
+    public void post() {
+        RealmNotifier notifier = new AndroidRealmNotifier(null, capabilitiesCanDeliver);
+        notifier.post(new Runnable() {
+            @Override
+            public void run() {
+                looperThread.testComplete();
+            }
+        });
+    }
+
+    @Test
+    @RunTestInLooperThread
+    public void postAtFrontOfQueue() {
+        final RealmNotifier notifier = new AndroidRealmNotifier(null, capabilitiesCanDeliver);
+        notifier.post(new Runnable() {
+            @Override
+            public void run() {
+                fail();
+            }
+        });
+        notifier.postAtFrontOfQueue(new Runnable() {
+            @Override
+            public void run() {
+                looperThread.testComplete();
+                notifier.close();
+            }
+        });
+    }
+
+    @Test
+    @RunTestInLooperThread
+    public void addChangeListener_byLocalChanges() {
+        SharedRealm sharedRealm = getSharedRealm(looperThread.realmConfiguration);
+        sharedRealm.realmNotifier.addChangeListener(sharedRealm, new RealmChangeListener<SharedRealm>() {
+            @Override
+            public void onChange(SharedRealm sharedRealm) {
+                // Transaction has been committed in core, but commitTransaction hasn't returned in java.
+                // Need a flag in java.
```

Comment:
Not really. I believe in #3975 it is caused by the Realm instance has not received change notification...

But I will solve this one here by setting a flag in between `beginTransaction` and `commitTransaction`.